portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -25,25 +33,110 @@
 
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private List<String> quotes;
-
-  @Override
-  public void init() {
-    quotes = new ArrayList<>(4);
-    quotes.add("Hope you are having a good one!");
-    quotes.add("How are you :)");
-    quotes.add("It is fantastic to see you!");
-    quotes.add("Thanks for vising my portfolio!");
-  }
+  private UserComment comment;
+  private static final int DEFAULT_MAX_COMMENTS = 10;
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Convert quotes object to JSON format.
+    // Query comment history from Datastore.
+    Query commentHistoryQuery = new Query("UserComment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery commentHistory = datastore.prepare(commentHistoryQuery);
+    int totalComments = commentHistory.countEntities(FetchOptions.Builder.withDefaults());
+
+    // Obtain maximum number of comments to display. If the number is invalid, take on the default value.
+    int maxN;
+    try {
+      maxN = Integer.parseInt(request.getParameter("maxN"));
+      if (maxN < 0) {
+        maxN = DEFAULT_MAX_COMMENTS;
+      }
+    } catch (NumberFormatException e) {
+      maxN = DEFAULT_MAX_COMMENTS;
+    }
+
+    // Convert comment Datastore entity into UserComment objects.
+    List<UserComment> comments = new ArrayList<>(maxN);
+    int commentCounter = 0;
+    for (Entity commentEntity : commentHistory.asIterable()) {
+      if (commentCounter == maxN) {
+        break;
+      }
+      String message = (String) commentEntity.getProperty("message");
+      String name = (String) commentEntity.getProperty("name");
+      String email = (String) commentEntity.getProperty("email");
+
+      UserComment commentItem = new UserComment(message, name, email);
+      comments.add(commentItem);
+      commentCounter++;
+    }
+
+    // Convert a history of comment objects, the total number of comments and default number, to JSON format.
+    Collection<Object> commentData = new ArrayList<>(2);
+    commentData.add(comments);
```

Comment:
Instead of using Collection<Object> can you create a new class and do Collection<CommentData>? That way when you're accessing the JSON objects, you should be able to do something like
 const commentDataJSON = await response.json();
  const comments = commentDataJSON.comment and so forth. 
This way you won't have to keep track of the order in which you added items to this Object

Reply:
I'll make the change. This will make the code a lot more extendable in the future. Thanks so much!

Reply:
Since the collection carries a list of comments and 2 numbers that describe the entire comment history, I defined a new class named CommentDataPackage to package together the 3 items into one object.